Lib/asyncio/sslproto.py

```diff
@@ -107,8 +107,11 @@ def close(self):
         protocol's connection_lost() method will (eventually) called
         with None as its argument.
         """
-        self._closed = True
-        self._ssl_protocol._start_shutdown()
+        if not self._closed:
+            self._closed = True
+            self._ssl_protocol._start_shutdown()
```

Comment:
So if we're taking this path through the code, the cycle remains unbroken?

Reply:
The cycle will be broken when the `_ssl_protocol` calls this `close` method again. When the shutdown is complete this method is called again and in that case `_closed` is true so we break the cycle.

Reply:
See https://github.com/python/cpython/blob/5871e19942fdcf83653924ae9a849941669c1143/Lib/asyncio/sslproto.py#L656-L664

The last line schedules to call this again when the whole shutdown is complete.

Reply:
How bizarre that it relies on calling `close()` twice. But everything about SSL is bizarre, so I'll merge this now. :-)

Reply:
"If it ain't broke, don't fix it" :)